guillotina/fields/annotation.py

```diff
@@ -215,3 +204,299 @@ class PatchBucketListRemove(PatchBucketListAppend):
             await existing.remove(context, value['bucket_index'], value['item_index'])
         except IndexError:
             raise ValueDeserializationError(self.field, value, 'Not valid index value')
+
+
+class BucketDictValue:
+    '''
+    metadata for managing bucket dict values
+    '''
+
+    def __init__(self, annotation_prefix='bucketdict-', bucket_len=1000):
+        self.buckets = [
+            {
+                'id': uuid.uuid4().hex,  # random gen
+                'len': 0,
+                'created': time.time(),
+                'start': None  # first one has no bound here
+            }
+        ]
+        self.annotation_prefix = annotation_prefix
+        self.bucket_len = bucket_len
+
+    def _find_bucket(self, key) -> typing.Tuple[int, dict]:
+        found = (0, self.buckets[0])
+        for idx, bucket in enumerate(self.buckets[1:]):
+            if key < bucket['start']:
+                # searched as far as we need to go
+                break
+            found = (idx + 1, bucket)
+        return found
+
+    def get_annotation_name(self, bucket_id: str) -> str:
+        return f'{self.annotation_prefix}{bucket_id}'
+
+    async def get_annotation(self, context, key=None,
+                             anno_id=None, create=True) -> typing.Optional[AnnotationData]:
+        if anno_id is None:
+            bidx, bucket = self._find_bucket(key)
+            annotation_name = self.get_annotation_name(bucket['id'])
+        else:
+            annotation_name = self.get_annotation_name(anno_id)
+
+        annotations_container = IAnnotations(context)
+        annotation = annotations_container.get(annotation_name, _default)
+        if annotation is _default:
+            annotation = await annotations_container.async_get(
+                annotation_name, _default)
+        if annotation is _default:
+            if not create:
+                return None
+            annotation = AnnotationData({
+                'keys': [],
+                'values': []
+            })
+            await annotations_container.async_set(annotation_name, annotation)
+        return annotation
+
+    async def assign(self, context, key, value):
+        annotation = await self.get_annotation(context, key)
+
+        if len(annotation['keys']) >= self.bucket_len:
+            # we need to split this bucket
+            bidx, bucket = self._find_bucket(key)
+            middle_idx = int(self.bucket_len / 2)
+            middle_key = annotation['keys'][middle_idx]
+            new_bucket = {
+                'id': uuid.uuid4().hex,
+                'start': middle_key,
+                'created': time.time(),
+            }
+            self.buckets.insert(bidx + 1, new_bucket)
+            new_annotation = await self.get_annotation(context, middle_key)
```

Comment:
New annotation register()

Reply:
Register is done in the `get_annotation` call